src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -1702,6 +1706,62 @@ namespace stan {
     }
     boost::phoenix::function<set_array_expr_type> set_array_expr_type_f;
 
+    void set_vec_expr_type::operator()(expression& e,
```

Comment:
This is a bad name for this function.  It should be called something like `infer_matrix_or_row_vec_expr_type` and indicate that it's trying to figure out the type on the fly based on the first element.  Same for other functions that are like this.